File: src/main/java/locus/api/mapper/BadBBCodeFixer.java
package locus.api.mapper;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

public class BadBBCodeFixer {
	private static final Map<Pattern, String> BBCODE_MAP = new HashMap<>();
	private static final int BBCODE_PATTERN_FLAGS = Pattern.CASE_INSENSITIVE | Pattern.DOTALL;

	static {
		BBCODE_MAP.put(Pattern.compile("\\[align=(.+?)\\](.+?)\\[/align=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<div align='$1'>$2</div>");
		BBCODE_MAP.put(Pattern.compile("\\[color=(.+?)\\](.+?)\\[/color=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<font color='$1'>$2</font>");
		BBCODE_MAP.put(Pattern.compile("\\[size=(.+?)\\](.+?)\\[/size=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<font size='$1'>$2</font>");
		BBCODE_MAP.put(Pattern.compile("\\[img=(.+?),(.+?)\\](.+?)\\[/img=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<img width='$1' height='$2' src='$3' />");
		BBCODE_MAP.put(Pattern.compile("\\[email=(.+?)\\](.+?)\\[/email=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<a href='mailto:$1'>$2</a>");
		BBCODE_MAP.put(Pattern.compile("\\[url=(.+?)\\](.+?)\\[/url=[^\\]]+\\]", BBCODE_PATTERN_FLAGS), "<a href='$1'>$2</a>");
	}

	public static String fix(String text) {
		if (text == null)
			return null;

		String html = text;

		for (Map.Entry<Pattern, String> entry: BBCODE_MAP.entrySet()) {
			html = entry.getKey().matcher(html).replaceAll(entry.getValue());
		}

		return html;
	}
}

File: app/src/main/java/locus/api/mapper/BadBBCodeFixer.kt
package locus.api.mapper

import java.util.regex.Pattern

internal object BadBBCodeFixer {
    private val BBCODE_PATTERN_FLAGS = Pattern.CASE_INSENSITIVE or Pattern.DOTALL
    private val BBCODE_MAP = mapOf(
        Pattern.compile("\\[align=(.+?)](.+?)\\[/align=[^]]+]", BBCODE_PATTERN_FLAGS) to "<div align='$1'>$2</div>",
        Pattern.compile("\\[color=(.+?)](.+?)\\[/color=[^]]+]", BBCODE_PATTERN_FLAGS) to "<font color='$1'>$2</font>",
        Pattern.compile("\\[size=(.+?)](.+?)\\[/size=[^]]+]", BBCODE_PATTERN_FLAGS) to "<font size='$1'>$2</font>",
        Pattern.compile("\\[img=(.+?),(.+?)](.+?)\\[/img=[^]]+]", BBCODE_PATTERN_FLAGS) to "<img width='$1' height='$2' src='$3' />",
        Pattern.compile("\\[email=(.+?)](.+?)\\[/email=[^]]+]", BBCODE_PATTERN_FLAGS) to "<a href='mailto:$1'>$2</a>",
        Pattern.compile("\\[url=(.+?)](.+?)\\[/url=[^]]+]", BBCODE_PATTERN_FLAGS) to "<a href='$1'>$2</a>"
    )

    @JvmStatic
    fun fix(text: String?): String? {
        if (text == null)
            return null

        var html: String = text

        for ((key, value) in BBCODE_MAP) {
            html = key.matcher(html).replaceAll(value)
        }

        return html
    }
}
